### PR TITLE
Implementing opinionated `Binary` fmt trait for floats

### PR DESCRIPTION
**Pull request by blyxyas**
This PR implements the [`Binary`](https://doc.rust-lang.org/core/fmt/trait.Binary.html) trait for `f32` and `f64`. Its documentation also clarifies why the style was choosen.

Instead of printing the simple binary representation (which doesn't make much sense, as IEEE 754 structures floats **much** differently) it differentiates the three sections that IEEE 754 stablishes, the normality and sign, the exponent and the significand.

So, the final output would be something like:

```rs
println!("{:b}", -43.123f64);
// 10 10001 110001111101111100111011011001000101101000011100100000000000000
```

Note that the exponent (first section) is little endian, while the mantissa (second section) is big endian. I thought about changing this, but I thought that the best way here is to just maintain the IEEE 754 standard and keep it this way.

### CODE REVIEW

**Comment by tgross35** [comment]
As-is this adds an insta-stable feature gate `binary_float_fmt` so it will need FCP

@rustbot label -T-libs +T-libs-api +needs-fcp

**Comment by blyxyas** [comment]
FCP like in Final Comment Period? Do I need to do something here?

**Comment by tgross35** [comment]
I believe that anything that changes public API needs to go through FCP - either on a RFC, or with a ACP, or on the PR itself. I don't think this is too controversial.

Nothing for you to do, somebody from libs-api just needs to look at it when they get the chance (currently assigned to Josh)

**Comment by m-ou-se** [comment]
Wouldn't it be reasonable to expect `{:b}` to just give the actual value in base 2?

E.g. `format!("{:b}", 9.75)` resulting in `"1001.11"`?

Perhaps it's arguable that that is much less common than the underlying binary representaiton. But if we implement `{:b}` as in this PR, then what about `{:x}`?

Representing a floating point value in base 16 is not uncommon. Several programming languages even support hexadecimal floating point literals.

So should `println!("{:x}", 9.75f32)"` print the value, like `9.c` (or `0x9.c`  or `0x9.cp0`), or should it print the ieee754 representation, like `411c0000` (or `0 82 1c0000` or …)?

File: library/core/src/fmt/float.rs [hunk 1]
```diff
@@ -223,8 +223,31 @@ macro_rules! floating {
...
+                    (bits << ($exponent + 3)),                    // Mantissa / Significand
+                )
+            }
+        }
     };
 }
 
-floating! { f32 }
-floating! { f64 }
+// Checking IEEE 754 docs. it seems like it should be 24 and 53, but the first two bits are reversed for normality and sign respectively.
```
Comment (m-ou-se) [inline library/core/src/fmt/float.rs:250]: > the first two bits are reversed for normality and sign respectively.

That is incorrect. There is no normality bit.

File: library/core/src/fmt/float.rs [hunk 2]
```diff
@@ -223,8 +223,31 @@ macro_rules! floating {
...
+        impl Binary for $ty {
+            fn fmt(&self, fmt: &mut Formatter<'_>) -> Result {
+                let bits = self.to_bits();
+                write!(
+                    fmt,
+                    "{:b} {:b} {:b}",
+                    (bits & $mask) >> ($significand + $exponent), // Sign / Normality
+                    (bits << 2 >> ($significand + 1)),            // Exponent
+                    (bits << ($exponent + 3)),                    // Mantissa / Significand
+                )
```
Comment (m-ou-se) [inline library/core/src/fmt/float.rs:244]: This does not work: this code prints the same for `0.0` and `2.0` (`"0 0 0"`).

**Comment by m-ou-se** [comment]
> Note that the exponent (first section) is little endian, while the mantissa (second section) is big endian.

What do you mean with that? Endianness is usually about bytes, not bits. All parts of a f32 and f64 have their most significant bit first.

**Comment by blyxyas** [comment]
Closing this as I don't feel motivated enough to keep working on it + I lost my original documents about IEEE 754 because I changed my distro.

Sorry for wasting your time, at least a possible future implementation has some kind of base.

**Pull request closed without merging**